core/services/keeper/registry_synchronizer_sync.go

```diff
@@ -68,44 +69,43 @@ func (rs *RegistrySynchronizer) fullSyncUpkeeps(reg Registry) error {
 			canceled = append(canceled, upkeepID)
 		}
 	}
-	if _, err := rs.orm.BatchDeleteUpkeepsForJob(rs.job.ID, canceled); err != nil {
+	if _, err := rs.orm.BatchDeleteUpkeepsForJob(ctx, rs.job.ID, canceled); err != nil {
 		return errors.Wrap(err, "failed to batch delete upkeeps from job")
 	}
 	return nil
 }
 
 // batchSyncUpkeepsOnRegistry syncs <syncUpkeepQueueSize> upkeeps at a time in parallel
 // for all the IDs within newUpkeeps slice
-func (rs *RegistrySynchronizer) batchSyncUpkeepsOnRegistry(reg Registry, newUpkeeps []big.Big) {
+func (rs *RegistrySynchronizer) batchSyncUpkeepsOnRegistry(ctx context.Context, reg Registry, newUpkeeps []big.Big) {
 	wg := sync.WaitGroup{}
-	wg.Add(len(newUpkeeps))
 	chSyncUpkeepQueue := make(chan struct{}, rs.syncUpkeepQueueSize)
 
 	done := func() { <-chSyncUpkeepQueue; wg.Done() }
 	for i := range newUpkeeps {
 		select {
-		case <-rs.chStop:
-			return
+		case <-ctx.Done():
 		case chSyncUpkeepQueue <- struct{}{}:
-			go rs.syncUpkeepWithCallback(&rs.registryWrapper, reg, &newUpkeeps[i], done)
+			wg.Add(1)
```

Comment:
Changed to incremental so we can still `Wait()` after context is closed for goroutines to return.